Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp

```diff
@@ -9,24 +9,167 @@
+    }
+
+    void TerrainClipmapGenerationPass::CompileResources(const AZ::RHI::FrameGraphCompileContext& context)
+    {
+        if (m_shaderResourceGroup != nullptr)
+        {
+            BindPassSrg(context, m_shaderResourceGroup);
+            m_shaderResourceGroup->Compile();
+        }
+        if (m_drawSrg != nullptr)
+        {
+            BindSrg(m_drawSrg->GetRHIShaderResourceGroup());
+            m_drawSrg->Compile();
+        }
+    }
+
+    void TerrainClipmapGenerationPass::BuildCommandListInternal(const AZ::RHI::FrameGraphExecuteContext& context)
+    {
+        //! Skip invoking the compute shader.
+        if (m_dispatchItem.m_arguments.m_direct.m_totalNumberOfThreadsX == 0 ||
+            m_dispatchItem.m_arguments.m_direct.m_totalNumberOfThreadsY == 0)
+        {
+            return;
+        }
+
+        AZ::RHI::CommandList* commandList = context.GetCommandList();
+
+        SetSrgsForDispatch(commandList);
+
+        commandList->Submit(m_dispatchItem);
+    }
+
+    void TerrainClipmapGenerationPass::OnShaderReinitialized([[maybe_unused]] const AZ::RPI::Shader& shader)
+    {
+        LoadShader();
+    }
+
+    void TerrainClipmapGenerationPass::OnShaderAssetReinitialized(
+        [[maybe_unused]] const AZ::Data::Asset<AZ::RPI::ShaderAsset>& shaderAsset)
+    {
+        LoadShader();
+    }
+
+    void TerrainClipmapGenerationPass::OnShaderVariantReinitialized([[maybe_unused]] const AZ::RPI::ShaderVariant& shaderVariant)
+    {
+        LoadShader();
+    }
+
+    void TerrainClipmapGenerationPass::LoadShader()
+    {
+        // Load ComputePassData...
+        const TerrainClipmapGenerationPassData* passData = AZ::RPI::PassUtils::GetPassData<TerrainClipmapGenerationPassData>(m_passDescriptor);
+        if (passData == nullptr)
+        {
+            AZ_Error(
+                "PassSystem", false, "[TerrainClipmapGenerationPass '%s']: Trying to construct without valid pass data!",
+                GetPathName().GetCStr());
+            return;
+        }
+
+        // Load Shader
+        AZ::Data::Asset<AZ::RPI::ShaderAsset> shaderAsset;
+        if (passData->m_shaderReference.m_assetId.IsValid())
+        {
+            shaderAsset = AZ::RPI::FindShaderAsset(passData->m_shaderReference.m_assetId, passData->m_shaderReference.m_filePath);
+        }
+
+        if (!shaderAsset.GetId().IsValid())
+        {
+            AZ_Error(
+                "PassSystem", false, "[TerrainClipmapGenerationPass '%s']: Failed to load shader '%s'!", GetPathName().GetCStr(),
+                passData->m_shaderReference.m_filePath.data());
+            return;
+        }
+
+        m_shader = AZ::RPI::Shader::FindOrCreate(shaderAsset);
+        if (m_shader == nullptr)
+        {
+            AZ_Error(
+                "PassSystem", false, "[TerrainClipmapGenerationPass '%s']: Failed to load shader '%s'!", GetPathName().GetCStr(),
+                passData->m_shaderReference.m_filePath.data());
+            return;
+        }
+
```

Comment:
There are utility functions in RPIUtils that could remove the need for most of this code.

Reply:
I'll try to clean this a bit.